### PR TITLE
Feature: Implement KZG commitment functionality and hints

### PR DESCRIPTION
**Pull request by whichqua**
This pull request adds a new hint related to the `KZG commitment` functionality, implementing KZG commitment generation and testing it. 
It includes the introduction of FFT and blob handling for KZG commitments.

## Type

- [x] feature
- [ ] bugfix
- [ ] dev (no functional changes, no API changes)
- [ ] fmt (formatting, renaming)
- [ ] build
- [ ] docs
- [ ] testing

## Related
#176 

Requires https://github.com/Moonsong-Labs/cairo-vm/pull/41

## Breaking changes?

- [ ] yes
- [x] no

### CODE REVIEW

File: crates/starknet-os/kzg/trusted_setup.txt
Comment (odesenfans) [inline crates/starknet-os/kzg/trusted_setup.txt:1]: What is this file and why is it necessary?

File: crates/starknet-os/src/hints/block_context.rs [hunk 1]
```diff
@@ -342,20 +342,23 @@ pub fn is_leaf(
...
 ) -> Result<(), HintError> {
-    // TODO: add use_kzg_da to BlockContext
-    let value = Felt252::ZERO;
+    let block_context = exec_scopes.get_ref::<BlockContext>(vars::scopes::BLOCK_CONTEXT)?;
 
-    println!("Warning: skipping kzg_da (use_kzg_da = false)");
+    let value = block_context.block_info().use_kzg_da;
+
+    if !value {
+        println!("Warning: skipping kzg_da (use_kzg_da = false)");
```
Comment (odesenfans) [inline crates/starknet-os/src/hints/block_context.rs:360]:
```suggestion
        log::warn!("skipping kzg_da (use_kzg_da = false)");
```

File: crates/starknet-os/src/hints/kzg.rs [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+const COMMITMENT_BYTES_LENGTH: usize = 48;
+const COMMITMENT_LOW_BIT_LENGTH: usize = (COMMITMENT_BYTES_LENGTH * 8) / 2;
+const BLOB_SUBGROUP_GENERATOR: &str = "39033254847818212395286706435128746857159659164139250548781411570340225835782";
+const BLS_PRIME: &str = "52435875175126190479447740508185965837690552500527637822603658699938581184513";
+
+fn is_power_of_2(n: usize) -> bool {
+    n > 0 && (n & (n - 1)) == 0
+}
+
+fn fft(coeffs: &[BigInt], generator: &BigInt, prime: &BigInt, bit_reversed: bool) -> Result<Vec<BigInt>, io::Error> {
```
Comment (odesenfans) [inline crates/starknet-os/src/hints/kzg.rs:32]: What is the original source of this code?

File: crates/starknet-os/src/hints/kzg.rs [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+const FIELD_ELEMENTS_PER_BLOB: usize = 4096;
+const COMMITMENT_BYTES_LENGTH: usize = 48;
+const COMMITMENT_LOW_BIT_LENGTH: usize = (COMMITMENT_BYTES_LENGTH * 8) / 2;
+const BLOB_SUBGROUP_GENERATOR: &str = "39033254847818212395286706435128746857159659164139250548781411570340225835782";
+const BLS_PRIME: &str = "52435875175126190479447740508185965837690552500527637822603658699938581184513";
+
+fn is_power_of_2(n: usize) -> bool {
+    n > 0 && (n & (n - 1)) == 0
+}
```
Comment (odesenfans) [inline crates/starknet-os/src/hints/kzg.rs:30]: This method is useless, `usize` has a `is_power_of_two()` method.

File: crates/starknet-os/src/hints/kzg.rs [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+    _exec_scopes: &mut ExecutionScopes,
+    ids_data: &HashMap<String, HintReference>,
+    ap_tracking: &ApTracking,
+    _constants: &HashMap<String, Felt252>,
+) -> Result<(), HintError> {
+    let state_updates_start = get_ptr_from_var_name("state_updates_start", vm, ids_data, ap_tracking)?;
+    // NOTE: This hint uses `ids.da_size` but the references provided are not valid
+    // NOTE: We end up using `state_updates_end` after referencing the cairo code
+    let state_updates_end = get_ptr_from_var_name("state_updates_end", vm, ids_data, ap_tracking)?;
+    let range: Vec<MaybeRelocatable> = vm
```
Comment (odesenfans) [inline crates/starknet-os/src/hints/kzg.rs:179]: Use `vm.get_integer_range`.

File: crates/starknet-os/src/hints/kzg.rs [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    )
+    segments.write_arg(
+        ids.kzg_commitment.address_,
+        execution_helper.polynomial_coefficients_to_kzg_commitment_callback(
+            execution_helper.da_segment
+        )
+    )"#
+};
+
+async fn write_kzg_commitment_address_async(
```
Comment (odesenfans) [inline crates/starknet-os/src/hints/kzg.rs:168]: There's nothing async in this function.

**Review by odesenfans — CHANGES_REQUESTED** [review]
Some simple comments, I need to review the code this is coming from in depth.

File: crates/starknet-os/src/hints/kzg.rs [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+
+        let group_mul_f_odd: Vec<BigInt> =
+            group.iter().take(f_odd.len()).zip(f_odd.iter()).map(|(g, f)| (g * f) % prime).collect();
+
+        let mut result = Vec::with_capacity(coeffs.len());
+        for i in 0..f_even.len() {
+            result.push((f_even[i].clone() + &group_mul_f_odd[i]) % prime);
+        }
+        for i in 0..f_even.len() {
+            // Ensure non-negative diff
```
Comment (odesenfans) [inline crates/starknet-os/src/hints/kzg.rs:51]:
```suggestion
            // Ensure non-negative diff by adding prime to the value before applying the modulo
```

File: crates/starknet-os/src/hints/kzg.rs [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+
+    if coeffs.is_empty() {
+        return Ok(vec![]);
+    }
+
+    let coeffs_len = coeffs.len();
+    assert!(coeffs_len.is_power_of_two(), "Length is not a power of two.");
+
+    let mut group = vec![BigInt::one()];
+    for _ in 1..coeffs_len {
```
Comment (odesenfans) [inline crates/starknet-os/src/hints/kzg.rs:67]: For similarity with the Python code.
```suggestion
    for _ in 0..(coeffs_len - 1) {
```

File: crates/starknet-os/src/hints/kzg.rs [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+        let last = group.last().ok_or(io::Error::other("Group is missing last element"))?;
```
Comment (odesenfans) [inline crates/starknet-os/src/hints/kzg.rs:68]: Not really an IO error... Define a `FftError` enum with `thiserror`.

File: crates/starknet-os/src/hints/kzg.rs [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+        group.push((last * generator) % prime);
+    }
+
+    let mut values = _fft(coeffs, &group, prime);
+
+    if bit_reversed {
+        let width = coeffs_len.trailing_zeros() as usize;
```
Comment (odesenfans) [inline crates/starknet-os/src/hints/kzg.rs:75]: The Python code uses `width = coeffs_len.bit_length() - 1`. `trailing_zeros` gives you the number of zeroes after the last set bit. Ex: tz(128) = 7, tz(127) = 0, etc. `coeffs_len` is guaranteed to be a power of two so I guess this is correct for this particular case.

Add a comment explaining the difference with the Python code and why this is correct, it definitely took me a minute to get it.

File: crates/starknet-os/src/hints/kzg.rs [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+                    .map_err(|_| io::Error::other("Invalid binary cast to usize"))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        values = perm.into_iter().map(|i| values[i].clone()).collect();
+    }
+
+    Ok(values)
+}
+
+fn split_commitment(num: BigInt) -> (BigInt, BigInt) {
```
Comment (odesenfans) [inline crates/starknet-os/src/hints/kzg.rs:89]: I don't see this code in `cairo-lang`, is this closed source? Please point me to the source code for reference.

**Review by odesenfans — CHANGES_REQUESTED** [review]
More comments on the FFT part, I need the source for the rest of the code to continue the review.

File: crates/starknet-os/kzg/trusted_setup.txt
Comment (odesenfans) [reply crates/starknet-os/kzg/trusted_setup.txt:1]: Okay, but how was this one generated? Is it okay to keep it in the repo? Can we even use this one for production use cases? I need more info.

File: crates/starknet-os/src/hints/kzg.rs (hunk 2 above)
Comment (notlesh) [reply crates/starknet-os/src/hints/kzg.rs:26]: Would be good to leave some kind of hint to help others find this. In other places we've copied the code comments from the fn (just below the line you linked to), which helps when `grep`ing.

File: crates/starknet-os/src/hints/kzg.rs [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+    execution_helper.store_da_segment(
+        da_segment=memory.get_range_as_ints(addr=ids.state_updates_start, size=ids.da_size)
+    )
+    segments.write_arg(
+        ids.kzg_commitment.address_,
+        execution_helper.polynomial_coefficients_to_kzg_commitment_callback(
+            execution_helper.da_segment
+        )
+    )"#
+};
```
Comment (notlesh) [inline crates/starknet-os/src/hints/kzg.rs:197]: I don't think this hint exists anymore, I think it was modified in `cairo-lang`

File: crates/starknet-os/src/hints/kzg.rs [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+    let da_size: usize = da_size.try_into().map_err(|_| HintError::BigintToU32Fail)?;
+
+    let da_segment: Vec<Felt252> =
+        vm.get_integer_range(state_updates_start, da_size)?.into_iter().map(|s| *s).collect();
+
+    // TODO: "BLOB_LENGTH" doesn't appear to be in scope, although it's clear from the OS code that it's
+    // 4096 let blob_length = get_integer_from_var_name(vars::ids::BLOB_LENGTH, vm, ids_data,
+    // ap_tracking)?.to_biguint(); let blob_length = get_constant(vars::ids::BLOB_LENGTH,
+    // constants)?.to_biguint(); let blob_length: usize = blob_length.try_into().map_err(|_|
+    // HintError::BigintToU32Fail)?;
```
Comment (notlesh) [inline crates/starknet-os/src/hints/kzg.rs:289]: Thanks, `cargo fmt`... this needs to be investigated (or, at a minimum this comment needs to be cleaned up).

File: crates/starknet-os/src/io/output.rs [hunk 13]
```diff
@@ -226,16 +226,16 @@ pub fn decode_output<I: Iterator<Item = Felt252>>(mut output_iter: I) -> Result<
     };
 
     if use_kzg_da {
-        // Skip KZG data.
-        _ = output_iter.by_ref().take(5);
+        // TODO: this needs an explanation
+        let _: Vec<_> = output_iter.by_ref().take(6).collect();
```
Comment (notlesh) [inline crates/starknet-os/src/io/output.rs:230]: Can someone explain this? Is `6` correct?

File: crates/starknet-os/src/hints/kzg.rs (hunk 11 above)
Comment (notlesh) [reply crates/starknet-os/src/hints/kzg.rs]: Removed in db1f6ac

File: crates/starknet-os/src/hints/kzg.rs (hunk 12 above)
Comment (notlesh) [reply crates/starknet-os/src/hints/kzg.rs]: Cleaned up the comment in 6f36912. I'm still not sure why we can't find `BLOB_LENGTH` as a constant, it looks like the same case as `MERKLE_HEIGHT` in the Cairo code which we fetch via `get_constant()` in SNOS.

**Review by odesenfans — CHANGES_REQUESTED** [review]
Add a block with KZG to the prove block itests.

File: crates/starknet-os/src/hints/os.rs [hunk 14]
```diff
@@ -32,14 +32,24 @@ pub const CONFIGURE_KZG_MANAGER: &str = indoc! {r#"__serialize_data_availability
...
+pub async fn configure_kzg_manager_async(
     _vm: &mut VirtualMachine,
     exec_scopes: &mut ExecutionScopes,
     _ids_data: &HashMap<String, HintReference>,
     _ap_tracking: &ApTracking,
     _constants: &HashMap<String, Felt252>,
 ) -> Result<(), HintError> {
     set_variable_in_root_exec_scope(exec_scopes, vars::scopes::SERIALIZE_DATA_AVAILABILITY_CREATE_PAGES, true);
-    // TODO: set kzg_manager
+
```
[the remark this answers is not in the document]
Comment (notlesh) [reply crates/starknet-os/src/hints/os.rs:51]: Isn't this done on line 41?

[the remark this answers is not in the document]
Comment (odesenfans) [reply crates/starknet-os/src/hints/os.rs:51]: Yes, you're right.

File: crates/starknet-os/src/hints/kzg.rs [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+    PCS: PerContractStorage + 'static,
+{
+    let state_updates_start = get_ptr_from_var_name(vars::ids::STATE_UPDATES_START, vm, ids_data, ap_tracking)?;
+    let da_size = get_integer_from_var_name(vars::ids::DA_SIZE, vm, ids_data, ap_tracking)?.to_biguint();
+    let da_size: usize = da_size.try_into().map_err(|_| HintError::BigintToU32Fail)?;
+
+    let da_segment: Vec<Felt252> =
+        vm.get_integer_range(state_updates_start, da_size)?.into_iter().map(|s| *s).collect();
+
+    // TODO: "BLOB_LENGTH" doesn't appear to be in scope either via get_integer_from_var_name()
```
[the remark this answers is not in the document]
Comment (notlesh) [reply crates/starknet-os/src/hints/kzg.rs]: Thanks! Fixed in 89f16f7

**Review by HermanObst — CHANGES_REQUESTED** [review]
@notlesh please add the integration test with the "blob" blocks that now are passing :)

File: crates/starknet-os/kzg/trusted_setup.txt
Comment (notlesh) [reply crates/starknet-os/kzg/trusted_setup.txt:1]: Where did it come from?

File: Cargo.toml [hunk 16]
```diff
@@ -36,6 +36,7 @@ cairo-lang-casm = { version = "=2.7.1" }
 cairo-type-derive = { version = "0.1.0", path = "crates/cairo-type-derive" }
 cairo-vm = { git = "https://github.com/Moonsong-Labs/cairo-vm", rev = "40ebe36e64ffa429ff18cba8ef0bd7d2c122ac12", features = ["extensive_hints", "cairo-1-hints"] }
 clap = { version = "4.5.4", features = ["derive"] }
+c-kzg = { version = "1" }
```
[the remark this answers is not in the document]
Comment (notlesh) [reply Cargo.toml]: Looks like this was addressed in 9ed7262e

File: crates/starknet-os/src/hints/kzg.rs [hunk 17]
```diff
@@ -0,0 +1,335 @@
...
+
+fn inner_fft(coeffs: &[BigInt], group: &[BigInt], prime: &BigInt) -> Vec<BigInt> {
+    if coeffs.len() == 1 {
+        return coeffs.to_vec();
+    }
+
+    // These calls involve a clone and a collect
+    // This not cheap and can possibly be improved with using `T: Iter<BigInt>`.
+    // It is tricky though as the method is very recursive and would require
+    // increasing the `recursive_limit` to crazy numbers
```
[the remark this answers is not in the document]
Comment (notlesh) [reply crates/starknet-os/src/hints/kzg.rs:61]: I don't have context on this, so I'm not sure what we need to do here. Maybe just leave a `// TODO`?

cc @whichqua

**Comment by notlesh** [comment]
> @notlesh please add the integration test with the "blob" blocks that now are passing :)

Done, I added two at random out of about a dozen that I found.

**Review by odesenfans — APPROVED** [review]

**Review by HermanObst — APPROVED** [review]

**Pull request merged**